File: src/lib/actions/conversations.ts
'use server';

import { and, desc, eq, inArray, sql } from 'drizzle-orm';
import { db, ensureMigrations } from '@/lib/db';
import { conversations, messages } from '@/lib/db/schema';
import { requireSession } from '@/lib/auth';
import type { Conversation, Message, Provider } from '@/lib/types';

async function init() {
  await requireSession();
  await ensureMigrations();
}

async function loadConversations(): Promise<Conversation[]> {
  const convs = await db.select().from(conversations).orderBy(desc(conversations.updatedAt));
  if (convs.length === 0) return [];
  const ids = convs.map((c) => c.id);
  const msgs = await db
    .select()
    .from(messages)
    .where(inArray(messages.conversationId, ids))
    .orderBy(messages.timestamp);
  const byConv = new Map<string, Message[]>();
  for (const m of msgs) {
    const arr = byConv.get(m.conversationId) ?? [];
    arr.push({
      id: m.id,
      role: m.role,
      content: m.content,
      timestamp: m.timestamp.getTime(),
      attachments: m.attachments ?? undefined,
    });
    byConv.set(m.conversationId, arr);
  }
  return convs.map((c) => ({
    id: c.id,
    title: c.title,
    provider: c.provider as Provider,
    model: c.model,
    createdAt: c.createdAt.getTime(),
    updatedAt: c.updatedAt.getTime(),
    messages: byConv.get(c.id) ?? [],
  }));
}

export async function listConversations(): Promise<Conversation[]> {
  await init();
  return loadConversations();
}

export async function createConversation(id: string, provider: Provider, model: string): Promise<void> {
  await init();
  const now = new Date();
  await db.insert(conversations).values({
    id,
    title: 'Nouvelle conversation',
    provider,
    model,
    createdAt: now,
    updatedAt: now,
  });
}

export async function addMessage(
  conversationId: string,
  message: Omit<Message, 'timestamp'>
): Promise<{ firstUserMessage: boolean }> {
  await init();
  const { id } = message;
  const now = new Date();

  // Determine if this is the first user message (drives title auto-rename).
  const existing = await db
    .select({ id: messages.id, role: messages.role })
    .from(messages)
    .where(eq(messages.conversationId, conversationId));
  const firstUserMessage =
    message.role === 'user' && existing.every((m) => m.role !== 'user');

  await db.insert(messages).values({
    id,
    conversationId,
    role: message.role,
    content: message.content,
    attachments: message.attachments ?? null,
    timestamp: now,
  });

  // Bump updatedAt; if first user message, also set the default title.
  const title = firstUserMessage
    ? message.content.slice(0, 40) + (message.content.length > 40 ? '...' : '')
    : undefined;
  await db
    .update(conversations)
    .set({ updatedAt: now, ...(title ? { title } : {}) })
    .where(eq(conversations.id, conversationId));

  return { firstUserMessage };
}

export async function updateMessage(
  conversationId: string,
  messageId: string,
  chunk: string,
  replace = false
): Promise<void> {
  await init();
  if (replace) {
    await db
      .update(messages)
      .set({ content: chunk })
      .where(and(eq(messages.id, messageId), eq(messages.conversationId, conversationId)));
  } else {
    await db
      .update(messages)
      .set({ content: sql`${messages.content} || ${chunk}` })
      .where(and(eq(messages.id, messageId), eq(messages.conversationId, conversationId)));
  }
  await db
    .update(conversations)
    .set({ updatedAt: new Date() })
    .where(eq(conversations.id, conversationId));
}

export async function renameConversation(id: string, title: string): Promise<void> {
  await init();
  if (!title) return;
  await db
    .update(conversations)
    .set({ title, updatedAt: new Date() })
    .where(eq(conversations.id, id));
}

export async function setConversationModel(
  id: string,
  provider: Provider,
  model: string
): Promise<void> {
  await init();
  await db
    .update(conversations)
    .set({ provider, model, updatedAt: new Date() })
    .where(eq(conversations.id, id));
}

export async function deleteConversation(id: string): Promise<void> {
  await init();
  await db.delete(conversations).where(eq(conversations.id, id));
}

File: apps/api/app/routers/conversations.py
"""Chat conversations + their messages — CRUD."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Annotated

from fastapi import APIRouter, Depends
from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import CurrentUser
from app.db.base import get_session
from app.db.models import Conversation, Message
from app.errors import NotFound
from app.schemas.conversations import (
    AddMessageRequest,
    AddMessageResponse,
    ConversationPayload,
    CreateConversationRequest,
    MessagePayload,
    UpdateConversationRequest,
    UpdateMessageRequest,
)
from app.utils import to_ms as _ms

router = APIRouter(prefix="/conversations", tags=["conversations"])

DBSession = Annotated[AsyncSession, Depends(get_session)]


async def _get_conv(session: AsyncSession, conv_id: str) -> Conversation:
    row = (
        await session.execute(select(Conversation).where(Conversation.id == conv_id))
    ).scalar_one_or_none()
    if row is None:
        raise NotFound(f"Conversation {conv_id} not found.")
    return row


# ─── Conversations CRUD ──────────────────────────────────────────────────────────────


@router.get("", response_model=list[ConversationPayload], response_model_by_alias=True)
async def list_conversations(user: CurrentUser, session: DBSession) -> list[ConversationPayload]:
    convs = (
        await session.execute(select(Conversation).order_by(Conversation.updated_at.desc()))
    ).scalars().all()
    if not convs:
        return []
    ids = [c.id for c in convs]
    msgs = (
        await session.execute(
            select(Message)
            .where(Message.conversation_id.in_(ids))
            .order_by(Message.timestamp.asc())
        )
    ).scalars().all()
    by_conv: dict[str, list[MessagePayload]] = {}
    for m in msgs:
        by_conv.setdefault(m.conversation_id, []).append(
            MessagePayload(
                id=m.id,
                role=m.role,  # type: ignore[arg-type]
                content=m.content,
                timestamp=_ms(m.timestamp),
                attachments=m.attachments,  # type: ignore[arg-type]
            )
        )
    return [
        ConversationPayload(
            id=c.id,
            title=c.title,
            provider=c.provider,
            model=c.model,
            messages=by_conv.get(c.id, []),
            created_at=_ms(c.created_at),
            updated_at=_ms(c.updated_at),
        )
        for c in convs
    ]


@router.post("", response_model=ConversationPayload, response_model_by_alias=True)
async def create_conversation(
    body: CreateConversationRequest, user: CurrentUser, session: DBSession
) -> ConversationPayload:
    row = Conversation(
        id=body.id,
        title="Nouvelle conversation",
        provider=body.provider,
        model=body.model,
    )
    session.add(row)
    await session.flush()
    return ConversationPayload(
        id=row.id,
        title=row.title,
        provider=row.provider,
        model=row.model,
        messages=[],
        created_at=_ms(row.created_at),
        updated_at=_ms(row.updated_at),
    )


@router.patch("/{conv_id}", response_model=ConversationPayload, response_model_by_alias=True)
async def update_conversation(
    conv_id: str,
    body: UpdateConversationRequest,
    user: CurrentUser,
    session: DBSession,
) -> ConversationPayload:
    row = await _get_conv(session, conv_id)
    if body.title is not None and body.title:
        row.title = body.title
    if body.provider is not None:
        row.provider = body.provider
    if body.model is not None:
        row.model = body.model
    row.updated_at = datetime.now(timezone.utc)
    await session.flush()
    return ConversationPayload(
        id=row.id,
        title=row.title,
        provider=row.provider,
        model=row.model,
        messages=[],
        created_at=_ms(row.created_at),
        updated_at=_ms(row.updated_at),
    )


@router.delete("/{conv_id}")
async def delete_conversation(
    conv_id: str, user: CurrentUser, session: DBSession
) -> dict[str, str]:
    # Postgres FK cascade deletes messages too.
    await session.execute(delete(Conversation).where(Conversation.id == conv_id))
    return {"status": "ok"}


# ─── Messages ────────────────────────────────────────────────────────────────────────


@router.post("/{conv_id}/messages", response_model=AddMessageResponse, response_model_by_alias=True)
async def add_message(
    conv_id: str,
    body: AddMessageRequest,
    user: CurrentUser,
    session: DBSession,
) -> AddMessageResponse:
    # Check if this is the first user message (auto-title)
    existing = (
        await session.execute(
            select(Message.role).where(Message.conversation_id == conv_id)
        )
    ).scalars().all()
    is_first_user = body.role == "user" and all(r != "user" for r in existing)

    now = datetime.now(timezone.utc)
    msg = Message(
        id=body.id,
        conversation_id=conv_id,
        role=body.role,
        content=body.content,
        attachments=[a.model_dump(by_alias=True) for a in body.attachments] if body.attachments else None,
        timestamp=now,
    )
    session.add(msg)

    conv = await _get_conv(session, conv_id)
    conv.updated_at = now
    if is_first_user:
        title = body.content[:40] + ("..." if len(body.content) > 40 else "")
        conv.title = title

    await session.flush()
    return AddMessageResponse(first_user_message=is_first_user)


@router.patch("/{conv_id}/messages/{msg_id}")
async def update_message(
    conv_id: str,
    msg_id: str,
    body: UpdateMessageRequest,
    user: CurrentUser,
    session: DBSession,
) -> dict[str, str]:
    msg = (
        await session.execute(
            select(Message).where(
                Message.id == msg_id, Message.conversation_id == conv_id
            )
        )
    ).scalar_one_or_none()
    if msg is None:
        raise NotFound(f"Message {msg_id} not found.")
    msg.content = body.chunk if body.replace else msg.content + body.chunk
    conv = await _get_conv(session, conv_id)
    conv.updated_at = datetime.now(timezone.utc)
    return {"status": "ok"}
